src/main/java/net/jodah/typetools/TypeResolver.java

```diff
@@ -479,33 +491,58 @@ public static Type resolveGenericType(Class<?> type, Type subType) {
     return genericType instanceof Class ? (Class<?>) genericType : Unknown.class;
   }
 
+  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableTypeMap) {
+    // Check for terminal cases to avoid allocating HashMap and trivial call.
+    if (genericType == null)
+      return null;
+    else if (genericType instanceof Class<?>)
+      return genericType;
+    else
+      return reify(genericType, typeVariableTypeMap, new HashMap<>());
+  }
+
   /**
    * Works like {@link #resolveRawClass(Type, Class, Class)} but does not stop at raw classes. Instead, traverses
    * referenced types.
    */
-  private static Type reify(final Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap) {
-    if (genericType == null) {
+  private static Type reify(Type genericType, final Map<TypeVariable<?>, Type> typeVariableMap, Map<Type, Type> cache) {
```

Comment:
We should probably add a javadoc comment mentioning what the `cache` is for.

Reply:
Well, it is not exposed publicly and just there to find cycles anyway, so I didn't comment it, but will do.

Reply:
Sure - I was just thinking for future maintainers / readers of the code.